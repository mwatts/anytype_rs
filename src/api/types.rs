use serde::{Deserialize, Serialize};

/// Color for tags and icons
#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, strum::Display)]
#[serde(rename_all = "lowercase")]
pub enum Color {
    Grey,
    Yellow,
    Orange,
    Red,
    Pink,
    Purple,
    Blue,
    Ice,
    Teal,
    Lime,
}

/// Generic API error response
#[derive(Debug, Deserialize)]
pub struct ApiErrorResponse {
    pub code: String,
    pub message: String,
    pub object: String,
    pub status: u32,
}

/// Pagination information
#[derive(Debug, Deserialize, Serialize)]
pub struct Pagination {
    pub has_more: bool,
    pub limit: usize,
    pub offset: usize,
    pub total: usize,
}

/// Icon format type
#[derive(Debug, Deserialize, Serialize, Clone)]
pub enum IconFormat {
    Emoji,
    File,
    Icon,
}

/// Icon enum that can be one of three types
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "format")]
pub enum Icon {
    #[serde(rename = "emoji")]
    Emoji { emoji: String },
    #[serde(rename = "file")]
    File { file: String },
    #[serde(rename = "icon")]
    Icon { color: Color, name: String },
}

/// Property information for types
#[derive(Debug, Deserialize, Serialize)]
pub struct TypeProperty {
    pub format: Format,
    pub id: String,
    pub key: String,
    pub name: String,
    pub object: String,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "snake_case")]
pub enum Format {
    Text,
    Number,
    Select,
    MultiSelect,
    Date,
    Files,
    Checkbox,
    Url,
    Email,
    Phone,
    Objects,
}

/// Type information
#[derive(Debug, Deserialize, Serialize)]
pub struct Type {
    pub archived: bool,
    pub icon: Option<Icon>,
    pub id: String,
    pub key: String,
    pub layout: Layout,
    pub name: String,
    pub object: String,
    pub plural_name: String,
    pub properties: Vec<TypeProperty>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "lowercase")]
pub enum Layout {
    Basic,
    Profile,
    Action,
    Note,
    Bookmark,
    Set,
    Collection,
    Participant,
}
