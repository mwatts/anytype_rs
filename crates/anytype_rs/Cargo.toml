[package]
name = "anytype_rs"
version.workspace = true
authors.workspace = true
description = "A Rust client library and CLI for the Anytype API"
repository.workspace = true
edition.workspace = true
license.workspace = true
keywords = ["anytype", "api", "client", "cli"]
categories = ["api-bindings", "command-line-interface"]
readme = "../../README.md"

[[bin]]
name = "anytype"
path = "src/cli/main.rs"

[dependencies]
anyhow.workspace = true
clap.workspace = true
dirs.workspace = true
gray_matter.workspace = true
reqwest.workspace = true
serde.workspace = true
serde_json.workspace = true
strum.workspace = true
thiserror.workspace = true
tokio.workspace = true
tracing.workspace = true
tracing-subscriber.workspace = true

[dev-dependencies]
insta.workspace = true
proptest.workspace = true

[lib]
name = "anytype_rs"
path = "src/lib.rs"
