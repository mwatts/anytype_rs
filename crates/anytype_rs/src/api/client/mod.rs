//! Anytype API client modules
//!
//! This module is organized to match the official API reference structure.

use crate::{error::Result, types::ApiErrorResponse};
use reqwest::{Client, Method, RequestBuilder};
use serde::{Serialize, de::DeserializeOwned};
use tracing::{debug, error};

// Include all module implementations
pub mod auth;
pub mod lists;
pub mod members;
pub mod objects;
pub mod properties;
pub mod search;
pub mod spaces;
pub mod tags;
pub mod templates;
pub mod types;

const DEFAULT_BASE_URL: &str = "http://localhost:31009";
const ANYTYPE_API_HEADER: &str = "Anytype-Version";
// TODO: Better support multiple API versions
const ANYTYPE_API_VERSION: &str = "2025-05-20";

/// Configuration for the Anytype client
#[derive(Debug, Clone)]
pub struct ClientConfig {
    pub base_url: String,
    pub timeout_seconds: u64,
    pub app_name: String,
}

impl Default for ClientConfig {
    fn default() -> Self {
        Self {
            base_url: DEFAULT_BASE_URL.to_string(),
            timeout_seconds: 30,
            app_name: "anytype_rs".to_string(),
        }
    }
}

/// Main client for interacting with the Anytype API
#[derive(Debug)]
pub struct AnytypeClient {
    pub(crate) http_client: Client,
    pub(crate) config: ClientConfig,
    pub(crate) api_key: Option<String>,
}

impl AnytypeClient {
    /// Create a new client with default configuration
    pub fn new() -> Result<Self> {
        Self::with_config(ClientConfig::default())
    }

    /// Create a new client with custom configuration
    pub fn with_config(config: ClientConfig) -> Result<Self> {
        let http_client = Client::builder()
            .timeout(std::time::Duration::from_secs(config.timeout_seconds))
            .build()?;

        Ok(Self {
            http_client,
            config,
            api_key: None,
        })
    }

    /// Set the API key for authenticated requests
    pub fn set_api_key(&mut self, api_key: String) {
        self.api_key = Some(api_key);
    }

    /// Get the current API key
    pub fn api_key(&self) -> Option<&str> {
        self.api_key.as_deref()
    }

    /// Make an authenticated GET request
    pub(crate) async fn get<T: DeserializeOwned>(&self, path: &str) -> Result<T> {
        let url = format!("{}{}", self.config.base_url, path);
        debug!("GET {}", url);

        let response = self
            .authenticated_request(Method::GET, &url)?
            .send()
            .await?;

        self.handle_response(response).await
    }

    /// Make an authenticated POST request with JSON body
    pub(crate) async fn post<T: DeserializeOwned, B: Serialize>(
        &self,
        path: &str,
        body: &B,
    ) -> Result<T> {
        let url = format!("{}{}", self.config.base_url, path);
        debug!("POST {}", url);

        let response = self
            .authenticated_request(Method::POST, &url)?
            .json(body)
            .send()
            .await?;

        self.handle_response(response).await
    }

    /// Make an authenticated PATCH request with JSON body
    pub(crate) async fn patch<T: DeserializeOwned, B: Serialize>(
        &self,
        path: &str,
        body: &B,
    ) -> Result<T> {
        let url = format!("{}{}", self.config.base_url, path);
        debug!("PATCH {}", url);

        let response = self
            .authenticated_request(Method::PATCH, &url)?
            .json(body)
            .send()
            .await?;

        self.handle_response(response).await
    }

    /// Make an authenticated DELETE request
    pub(crate) async fn delete<T: DeserializeOwned>(&self, path: &str) -> Result<T> {
        let url = format!("{}{}", self.config.base_url, path);
        debug!("DELETE {}", url);

        let response = self
            .authenticated_request(Method::DELETE, &url)?
            .send()
            .await?;

        self.handle_response(response).await
    }

    /// Make an unauthenticated POST request (for auth endpoints)
    pub(crate) async fn post_unauthenticated<T: DeserializeOwned, B: Serialize>(
        &self,
        path: &str,
        body: &B,
    ) -> Result<T> {
        let url = format!("{}{}", self.config.base_url, path);
        debug!("POST {} (unauthenticated)", url);

        let response = self
            .http_client
            .post(&url)
            .header(ANYTYPE_API_HEADER, ANYTYPE_API_VERSION)
            .json(body)
            .send()
            .await?;

        self.handle_response(response).await
    }

    /// Create an authenticated request builder (internal helper)
    fn authenticated_request(&self, method: Method, url: &str) -> Result<RequestBuilder> {
        let api_key = self
            .api_key
            .as_ref()
            .ok_or_else(|| crate::error::AnytypeError::Auth {
                message: "API key not set. Call set_api_key() first.".to_string(),
            })?;

        let builder = match method {
            Method::GET => self.http_client.get(url),
            Method::POST => self.http_client.post(url),
            Method::PATCH => self.http_client.patch(url),
            Method::DELETE => self.http_client.delete(url),
            _ => {
                return Err(crate::error::AnytypeError::Api {
                    message: format!("Unsupported HTTP method: {method}"),
                });
            }
        };

        Ok(builder
            .header(ANYTYPE_API_HEADER, ANYTYPE_API_VERSION)
            .bearer_auth(api_key))
    }

    /// Handle HTTP response and deserialize JSON
    pub(crate) async fn handle_response<T: DeserializeOwned>(
        &self,
        response: reqwest::Response,
    ) -> Result<T> {
        let status = response.status();
        let url = response.url().clone();

        debug!("Response status: {} for {}", status, url);

        if status.is_success() {
            // Get the response text first for debugging
            let response_text =
                response
                    .text()
                    .await
                    .map_err(|e| crate::error::AnytypeError::InvalidResponse {
                        message: format!("Failed to read response body: {e}"),
                    })?;

            debug!("Response body: {response_text}");

            let response = serde_json::from_str::<T>(&response_text);

            match response {
                Ok(data) => Ok(data),
                Err(e) => {
                    error!("Failed to deserialize response: {}", e);
                    error!("Expected type: {}", std::any::type_name::<T>());
                    error!("Response body was: {}", response_text);
                    Err(crate::error::AnytypeError::InvalidResponse {
                        message: format!(
                            "Failed to parse JSON response: {}. Expected type: {}",
                            e,
                            std::any::type_name::<T>()
                        ),
                    })
                }
            }
        } else {
            let response = response.json::<ApiErrorResponse>().await;
            error!("API error {}", status);

            match response {
                Ok(error) => {
                    let message = error.message;

                    if status == 401 || status == 403 {
                        Err(crate::error::AnytypeError::Auth { message })
                    } else {
                        Err(crate::error::AnytypeError::Api { message })
                    }
                }
                Err(e) => Err(crate::error::AnytypeError::Api {
                    message: format!("HTTP {status} - {e}"),
                }),
            }
        }
    }
}
