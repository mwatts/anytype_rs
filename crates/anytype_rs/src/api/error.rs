use thiserror::Error;

#[derive(Error, Debug)]
pub enum AnytypeError {
    #[error("HTTP request failed: {source}")]
    Http {
        #[from]
        source: reqwest::Error,
    },

    #[error("Authentication failed: {message}")]
    Auth { message: String },

    #[error("API error: {message}")]
    Api { message: String },

    #[error("Serialization error: {source}")]
    Serialization {
        #[from]
        source: serde_json::Error,
    },

    #[error("Invalid response: {message}")]
    InvalidResponse { message: String },
}

pub type Result<T> = std::result::Result<T, AnytypeError>;
