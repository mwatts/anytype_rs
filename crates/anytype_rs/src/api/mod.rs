//! # Anytype API Client
//!
//! A Rust library for interacting with your local Anytype application API.
//!
//! ## Features
//!
//! - Authentication via challenge-response mechanism with local Anytype app
//! - JWT Bearer token support
//! - Full CRUD operations for spaces and objects
//! - Search functionality
//! - Template, type, property, and tag management
//! - Async/await support with tokio
//! - Comprehensive error handling

pub mod client;
pub mod error;
pub mod types;

pub use client::{AnytypeClient, ClientConfig};
pub use error::{AnytypeError, Result};
pub use types::*;

// Re-export types from client modules for convenience
pub use client::auth::{
    CreateApiKeyRequest, CreateApiKeyResponse, CreateChallengeRequest, CreateChallengeResponse,
};
pub use client::lists::{
    AddListObjectsRequest, AddListObjectsResponse, GetListObjectsResponse, GetListViewsResponse,
    ListObject, ListObjectType, ListViewData, ListViewFilter, ListViewSort, ObjectTypeProperty,
    RemoveListObjectsResponse,
};
pub use client::members::{
    GetMemberResponse, ListMembersResponse, Member, MemberRole, MemberStatus,
};
pub use client::objects::{
    CreateObjectRequest, CreateObjectResponse, DeleteObjectResponse, ListObjectsResponse, Object,
    UpdateObjectRequest, UpdateObjectResponse,
};
pub use client::properties::{
    CreatePropertyRequest, CreatePropertyResponse, DeletePropertyResponse, GetPropertyResponse,
    ListPropertiesResponse, Property, UpdatePropertyRequest, UpdatePropertyResponse,
};
pub use client::search::{
    SearchObject, SearchRequest, SearchResponse, SearchSpaceRequest, Sort, SortDirection,
    SortProperty,
};
pub use client::spaces::{
    CreateSpaceRequest, CreateSpaceResponse, ListSpacesResponse, Space, UpdateSpaceRequest,
    UpdateSpaceResponse,
};
pub use client::tags::{
    CreateTagRequest, CreateTagResponse, DeleteTagResponse, GetTagResponse, ListTagsResponse, Tag,
    UpdateTagRequest, UpdateTagResponse,
};
pub use client::templates::{GetTemplateResponse, ListTemplatesResponse, ObjectType, Template};
pub use client::types::{
    CreateTypeProperty, CreateTypeRequest, CreateTypeResponse, DeleteTypeResponse, GetTypeResponse,
    Layout, ListTypesResponse, PropertyFormat, Type, TypeProperty, UpdateTypeRequest,
    UpdateTypeResponse,
};
pub use types::{Icon, IconFormat};
