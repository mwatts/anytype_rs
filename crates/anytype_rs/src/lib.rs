//! # Anytype Rust Client and CLI
//!
//! A Rust library and command-line interface for interacting with your local Anytype application API.
//!
//! ## Features
//!
//! - Authentication via challenge-response mechanism with local Anytype app
//! - JWT Bearer token support
//! - Full CRUD operations for spaces and objects
//! - Search functionality
//! - Template, type, property, and tag management
//! - Async/await support with tokio
//! - Comprehensive error handling
//! - Command-line interface
//!
//! ## Library Usage
//!
//! ```rust,no_run
//! use anytype_rs::{AnytypeClient, Result};
//!
//! #[tokio::main]
//! async fn main() -> Result<()> {
//!     // Connect to local Anytype app (http://localhost:31009)
//!     let mut client = AnytypeClient::new()?;
//!     
//!     // Authenticate (you'll need to implement the challenge flow)
//!     client.set_api_key("your-jwt-token".to_string());
//!     
//!     // List spaces from your local Anytype
//!     let spaces = client.list_spaces().await?;
//!     println!("Found {} spaces", spaces.len());
//!     
//!     Ok(())
//! }
//! ```
//!
//! ## CLI Usage
//!
//! The crate also provides a command-line interface:
//!
//! ```bash
//! # Install the CLI
//! cargo install anytype_rs
//!
//! # Use the CLI
//! anytype spaces list
//! anytype search "my query"
//! anytype types list <space_id>
//! ```

pub mod api;

// Re-export the main API types for convenience
pub use api::*;
