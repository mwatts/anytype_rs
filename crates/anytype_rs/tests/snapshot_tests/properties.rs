//! Snapshot tests for properties module types

use anytype_rs::api::{CreatePropertyRequest, Property, PropertyFormat, UpdatePropertyRequest};

#[test]
fn test_property_format_serialization() {
    insta::assert_json_snapshot!("property_format_text", PropertyFormat::Text);
    insta::assert_json_snapshot!("property_format_number", PropertyFormat::Number);
    insta::assert_json_snapshot!("property_format_select", PropertyFormat::Select);
    insta::assert_json_snapshot!("property_format_multi_select", PropertyFormat::MultiSelect);
    insta::assert_json_snapshot!("property_format_date", PropertyFormat::Date);
    insta::assert_json_snapshot!("property_format_files", PropertyFormat::Files);
    insta::assert_json_snapshot!("property_format_checkbox", PropertyFormat::Checkbox);
    insta::assert_json_snapshot!("property_format_url", PropertyFormat::Url);
    insta::assert_json_snapshot!("property_format_email", PropertyFormat::Email);
    insta::assert_json_snapshot!("property_format_phone", PropertyFormat::Phone);
    insta::assert_json_snapshot!("property_format_objects", PropertyFormat::Objects);
}

#[test]
fn test_property_serialization() {
    let property = Property {
        format: "text".to_string(),
        id: "prop123".to_string(),
        key: "title".to_string(),
        name: "Title".to_string(),
        object: "property".to_string(),
    };
    insta::assert_json_snapshot!("property_text", property);

    let property_date = Property {
        format: "date".to_string(),
        id: "prop456".to_string(),
        key: "created_at".to_string(),
        name: "Created At".to_string(),
        object: "property".to_string(),
    };
    insta::assert_json_snapshot!("property_date", property_date);
}

#[test]
fn test_create_property_request_serialization() {
    let request = CreatePropertyRequest {
        name: "New Property".to_string(),
        format: PropertyFormat::Text,
        key: None,
    };
    insta::assert_json_snapshot!("create_property_request_text", request);

    let request_select = CreatePropertyRequest {
        name: "Status".to_string(),
        format: PropertyFormat::Select,
        key: None,
    };
    insta::assert_json_snapshot!("create_property_request_select", request_select);
}

#[test]
fn test_update_property_request_serialization() {
    let request = UpdatePropertyRequest {
        name: "Updated Property".to_string(),
        key: Some("updated_prop".to_string()),
    };
    insta::assert_json_snapshot!("update_property_request", request);
}
