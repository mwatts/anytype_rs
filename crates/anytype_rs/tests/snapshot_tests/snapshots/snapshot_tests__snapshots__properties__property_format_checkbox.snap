---
source: tests/snapshot_tests/properties.rs
expression: "PropertyFormat::Checkbox"
---
"checkbox"
